### PR TITLE
[MI-2606] Add feature to add filters to bulk subscriptions

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

File: server/plugin/client.go [hunk 1]
```diff
@@ -329,3 +334,21 @@ func (c *client) SearchFilterValuesInServiceNow(searchTerm, limit, offset, reque
 
 	return assignmentGroups.Result, statusCode, nil
 }
+
+func (c *client) GetTableFieldsFromServiceNow(table string) ([]*serializer.ServiceNowTableFields, int, error) {
+	fields := &serializer.ServiceNowTableFieldsResult{}
+	queryParams := url.Values{
```
Comment (Kshitij-Katiyar) [inline server/plugin/client.go:340]: make this one line

**Review by Kshitij-Katiyar — APPROVED** [review]

File: webapp/src/containers/addOrEditSubscriptions/subComponents/styles.scss [hunk 2]
```diff
@@ -55,4 +55,20 @@
...
+
+  .filters-panel {
+    min-height: 247px;
+
+    &__label {
+      color: rgba(var(--center-channel-color-rgb), 0.8);
+      margin: 12px 0 12px 25px;
+    }
+
+    &__auto-suggest {
```
Comment (manojmalik20) [inline webapp/src/containers/addOrEditSubscriptions/subComponents/styles.scss:67]: Please get reviewed by a FE dev.

File: webapp/src/services/index.ts [hunk 3]
```diff
@@ -159,6 +159,22 @@ const pluginApi = createApi({
...
         }),
+        [Constants.pluginApiServiceConfigs.getFilterData.apiServiceName]: builder.query<FieldsFilterData[], SearchFilterItemsParams>({
+            query: ({search, filter, perPage}) => ({
+                headers: {[Constants.HeaderCSRFToken]: Cookies.get(Constants.MMCSRF)},
+                url: `${Constants.pluginApiServiceConfigs.getFilterData.path}/${filter}`,
+                method: Constants.pluginApiServiceConfigs.getFilterData.method,
+                params: {search, perPage: perPage || Constants.DefaultPerPageParam},
+            }),
+        }),
+        [Constants.pluginApiServiceConfigs.getTableFeilds.apiServiceName]: builder.query<void, string>({
```
Comment (manojmalik20) [inline webapp/src/services/index.ts:170]:
```suggestion
        [Constants.pluginApiServiceConfigs.getTableFields.apiServiceName]: builder.query<void, string>({
```

File: webapp/src/types/common/service.d.ts [hunk 4]
```diff
@@ -20,7 +20,9 @@ type ApiServiceName =
     'searchItems' |
     'getUsers' |
     'createIncident' |
-    'getIncidentFeilds';
+    'getIncidentFeilds' |
```
Comment (manojmalik20) [inline webapp/src/types/common/service.d.ts:23]:
```suggestion
    'getIncidentFields' |
```

File: server/plugin/api_test.go [hunk 5]
```diff
@@ -1227,8 +1230,9 @@ func TestCreateSubscription(t *testing.T) {
 		"failed to create subscription": {
 			RequestBody: fmt.Sprintf(`{
 				"user_id": "%s",
+				"type": "%s",
 				"channel_id": "%s"
-			  	}`, testutils.GetID(), testutils.GetChannelID()),
+			  	}`, testutils.GetID(), constants.SubscriptionTypeRecord, testutils.GetChannelID()),
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1235]: You can create a testutil for getting this request body.

File: server/constants/constants.go [hunk 6]
```diff
@@ -115,6 +118,8 @@ const (
 	APIErrorCreateIncident               = "Error in creating the incident"
 	APIErrorSearchingCatalogItems        = "Error in searching for catalog items in ServiceNow"
 	ServiceNowAPIErrorURINotPresent      = "Requested URI does not represent any resource"
+	APIErrorIDAccessTable                = "unauthorized_to_access_table"
+	APIErrorAccessTable                  = "unauthorized to access table"
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/constants/constants.go:123]: we were using this to check error returned from API, added a new constant here

File: webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+            } else {
+                payload.filter = SupportedFilters.SERVICE;
+            }
+
+            setSearchItemsPayload(payload);
+            makeApiRequestWithCompletionStatus(Constants.pluginApiServiceConfigs.getFilterData.apiServiceName, payload);
+        }
+    };
+
+    const mapRequestsToSuggestions = (data: FieldsFilterData[]): Array<Record<string, string>> => data.map((d) => ({
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx]: updated the name

File: webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        if (currentValue) {
+            if (currentValue.length >= Constants.DefaultCharThresholdToShowSuggestions) {
+                debouncedGetSuggestions({searchFor: currentValue, type: SupportedFilters.ASSIGNMENT_GROUP});
+            }
+        }
+    };
+
+    const handleServiceInputChange = (currentValue: string) => {
+        setServiceAutoSuggestValue(currentValue);
+        setFiltersValue(SupportedFilters.SERVICE, null);
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx:96]: no, on change of input value, we are resetting irrespective of input length

File: webapp/src/containers/addOrEditSubscriptions/subComponents/index.tsx [hunk 9]
```diff
@@ -218,13 +237,17 @@ const AddOrEditSubscription = ({open, close, subscriptionData}: AddOrEditSubscri
         setRecordType(null);
         setSubscriptionEvents([]);
         setEditSubscriptionData(null);
+        setFilters([]);
+        setResetFiltersPanelStates(true);
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply webapp/src/containers/addOrEditSubscriptions/subComponents/index.tsx:241]: this is used in some other panels as well like search records panel

**Review by sooraj-shukla — APPROVED** [review]

File: webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (manojmalik20) [reply webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx:96]: But why?

File: webapp/src/services/index.ts [hunk 10]
```diff
@@ -167,12 +167,11 @@ const pluginApi = createApi({
...
         }),
-        [Constants.pluginApiServiceConfigs.getTableFeilds.apiServiceName]: builder.query<void, string>({
+        [Constants.pluginApiServiceConfigs.getTableFields.apiServiceName]: builder.query<void, string>({
             query: (table) => ({
                 headers: {[Constants.HeaderCSRFToken]: Cookies.get(Constants.MMCSRF)},
-                url: Constants.pluginApiServiceConfigs.getTableFeilds.path,
-                method: Constants.pluginApiServiceConfigs.getTableFeilds.method,
-                params: {table},
+                url: `${Constants.pluginApiServiceConfigs.getTableFields.path}/${table}`,
+                method: Constants.pluginApiServiceConfigs.getTableFields.method,
```
Comment (manojmalik20) [inline webapp/src/services/index.ts:174]: Why did you remove the `params` field?

File: server/plugin/api_test.go [hunk 11]
```diff
@@ -1065,7 +1065,7 @@ func TestCreateSubscription(t *testing.T) {
 				"user_id": "%s",
 				"type": "%s",
 				"channel_id": "%s"
-			  	}`, testutils.GetID(), constants.SubscriptionTypeRecord, testutils.GetChannelID()),
+			  	}`, testutils.GetID(), testutils.GetRecordTypeSubscription(), testutils.GetChannelID()),
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1068]: I meant, you should create a testutil function for getting the whole request body, not just the subscription type.

File: webapp/src/services/index.ts (hunk 10 above)
Comment (ayusht2810) [reply webapp/src/services/index.ts:174]: discussed

File: webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ayusht2810) [reply webapp/src/containers/addOrEditSubscriptions/subComponents/filtersPanel.tsx:96]: discussed

**Review by manojmalik20 — APPROVED** [review]
